proc: Add logical expressions in comparisons

akmorrison opened the pull request:
So far this supports and, or, and parentheses. And is taken as having
higher precedence, so "x and y or z" is parsed as "(x and y) or z".

There are three terminal symbols for and and or symbols, "AND, "and",
and "&&". I'm not sure if that's the desired behaviour.

Addresses #61

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/waterloo-rocketry/topside/66)
<!-- Reviewable:end -->

akmorrison committed:
proc: Add logical expressions in comparisons

So far this supports and, or, and parentheses. And is taken as having
higher precedence, so "x and y or z" is parsed as "(x and y) or z".

There are three terminal symbols for and and or symbols, "AND, "and",
and "&&". I'm not sure if that's the desired behaviour.

akmorrison commented on topside/procedures/proclang.py:
```diff
@@ -90,14 +102,35 @@ def ge(self, data):
+            return top.Or(data)
+
+    def comparison_and(self, data):
+        print(data)
```
This shouldn't be there, whoops

jacobdeery reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/topside/66)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @akmorrison and @jacobdeery)

___
*[topside/procedures/proclang.py, line 36 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhkeFQDq5L7ExRv9uQ:-MHhkeFQDq5L7ExRv9uR:b1buor2) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/proclang.py#L36)):*
> ```
> comparison: comparison_and ( logic_or comparison_and )+
>     | comparison_and
> ```
any reason this can't be `comparison_and ( logic_or comparison_and )*`? (same for next rule)
___
*[topside/procedures/proclang.py, line 42 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhlMQuDjNFvn6L630T:-MHhlMQuDjNFvn6L630U:b-tj8118) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/proclang.py#L42)):*
> ```
> comp_direct: node operator value
>     |   "(" comparison ")"
> ```
hmm, in this form we can't write things like `500s and p1 < 500`; can we modify this so that it can accept any type of `condition`? will make it easier to expand in the future with new and advanced condition types.
___
*[topside/procedures/proclang.py, line 122 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhml9z0pzRwu8aoTX2:-MHhml9z0pzRwu8aoTX3:b-kjdp9q) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/proclang.py#L122)):*
> ```Python
>             return data[0]
> 
>     def comparison(self, data):
> ```

Can you docstring these? Doesn't need to be super detailed, just see the above functions for reference
___
*[topside/procedures/tests/test_proclang.py, line 274 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhlpsW3MdfkVItYVRR:-MHhlpsW3MdfkVItYVRS:bmll2x4) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/tests/test_proclang.py#L274)):*
> ```Python
>     assert suite == expected_suite
> 
> def test_parse_steps_with_complex_comparisons():
> ```

Can we either add some tests that use the alternate forms (`&&`, `AND`) or just tweak some of the statements in this test so that we have coverage on that as well?


<!-- Sent from Reviewable.io -->

akmorrison committed:
proc: Review changes for proclang logical operators

Changes in this commit:
        - Add docstrings for parse tree transformers
        - Use more terse rules for expressions
        - Add tests for different operator forms ("and" vs "&&")
        - Allow arbitrary logical expressions involving either pressure
          measurements or wait times

akmorrison reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/topside/66)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @jacobdeery)

___
*[topside/procedures/proclang.py, line 36 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhkeFQDq5L7ExRv9uQ:-MHhqQnVBlDKlRhJtb2p:boo6xsm) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/proclang.py#L36)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> comparison: comparison_and ( logic_or comparison_and )+
>     | comparison_and
> ```
any reason this can't be `comparison_and ( logic_or comparison_and )*`? (same for next rule)
</blockquote></details>

Nope, will fix
___
*[topside/procedures/proclang.py, line 42 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhlMQuDjNFvn6L630T:-MHhqYxA7AnfLtap730b:b-wcmsch) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/proclang.py#L42)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> comp_direct: node operator value
>     |   "(" comparison ")"
> ```
hmm, in this form we can't write things like `500s and p1 < 500`; can we modify this so that it can accept any type of `condition`? will make it easier to expand in the future with new and advanced condition types.
</blockquote></details>

Sure, I'll rejigger it so that we can have arbitrary conditions combined with arbitrary logical operators. We don't want full expression syntax though, right? Like, `p1 < 500 < p2` shouldn't be legal
___
*[topside/procedures/proclang.py, line 122 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhml9z0pzRwu8aoTX2:-MHhqlqe1ROU4Htmnw_3:bb74ce2) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/proclang.py#L122)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Can you docstring these? Doesn't need to be super detailed, just see the above functions for reference
</blockquote></details>

Sure
___
*[topside/procedures/tests/test_proclang.py, line 274 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhlpsW3MdfkVItYVRR:-MHhqqPtBYGZFrwYOeAy:brxdhu7) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/tests/test_proclang.py#L274)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Can we either add some tests that use the alternate forms (`&&`, `AND`) or just tweak some of the statements in this test so that we have coverage on that as well?
</blockquote></details>

Wilco. I'll also add in some tests for time delays "500s or p1 < 12" or something


<!-- Sent from Reviewable.io -->

jacobdeery reviewed (commented):
Cool! Just some minor formatting comments and then should be good to go

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-:-MHiF0THAHRz3CVa0OfR:b-1j0kst)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @akmorrison and @jacobdeery)

___
*[topside/procedures/proclang.py, line 42 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhlMQuDjNFvn6L630T:-MHiDkpB99moVjRPZBIa:b-prtkli) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/proclang.py#L42)):*
<details><summary><i>Previously, akmorrison (Aaron Morrison) wrote…</i></summary><blockquote>

Sure, I'll rejigger it so that we can have arbitrary conditions combined with arbitrary logical operators. We don't want full expression syntax though, right? Like, `p1 < 500 < p2` shouldn't be legal
</blockquote></details>

Agreed, this looks good
___
*[topside/procedures/proclang.py, line 129 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiE0z9CvRAy68wpsWN:-MHiE0z9CvRAy68wpsWO:ba33ha2) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/proclang.py#L129)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
>         elif len(data) == 1 and type(data[0]) == top.WaitUntil:
>             return data[0]
>         elif len(data) == 1:
>             # Assume that this is a parenthesized expression
>             return data[0]
> ```
</details>

You can collapse these into one branch if you want, but you don't need to if you prefer keeping it this way for readability
___
*[topside/procedures/tests/test_proclang.py, line 274 at r1](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHhlpsW3MdfkVItYVRR:-MHiErOo8SORbYEfUoN6:bsrfkwe) ([raw file](https://github.com/waterloo-rocketry/topside/blob/86c9e9f1f6cd5217e3a1f896f19b1c5cfce5a869/topside/procedures/tests/test_proclang.py#L274)):*
<details><summary><i>Previously, akmorrison (Aaron Morrison) wrote…</i></summary><blockquote>

Wilco. I'll also add in some tests for time delays "500s or p1 < 12" or something
</blockquote></details>

Looks good
___
*[topside/procedures/tests/test_proclang.py, line 287 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiEZ8o1JpDo1QScX2G:-MHiEZ8o1JpDo1QScX2H:bt63ubz) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/tests/test_proclang.py#L287)):*
> ```
> compAnd
> ```
Nit: `snake_case`
___
*[topside/procedures/tests/test_proclang.py, line 359 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiEOAD6AYN0Ebw-K-j:-MHiEOAD6AYN0Ebw-K-k:b-42hdsu) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/tests/test_proclang.py#L359)):*
> ```
> set vent_valve to open
> ```
we really want to open the vent valve
___
*[topside/procedures/tests/test_proclang.py, line 363 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiEchd601CuOvz572i:-MHiEchd601CuOvz572j:b-3u2apl) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/tests/test_proclang.py#L363)):*
> ```
> compAnd
> ```
Same
___
*[topside/procedures/tests/test_proclang.py, line 403 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiEV1ZEAxoLZmUcZim:-MHiEV1ZEAxoLZmUcZin:b-368zoh) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/tests/test_proclang.py#L403)):*
> ```
> (top.Or([top.Less('p1', 100), top.And([top.WaitUntil(1e6*500), top.Less('p2', 200)])]), top.Transition('main', '2'))
> ```
Nit: can you pull the `Or` outside of the suite for readability like you've done in previous tests?


<!-- Sent from Reviewable.io -->

akmorrison reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/topside/66)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @akmorrison and @jacobdeery)

___
*[topside/procedures/proclang.py, line 129 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiE0z9CvRAy68wpsWN:-MHiGXmm2G42FZRGf4iu:b-l6vfc5) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/proclang.py#L129)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
>         elif len(data) == 1 and type(data[0]) == top.WaitUntil:
>             return data[0]
>         elif len(data) == 1:
>             # Assume that this is a parenthesized expression
>             return data[0]
> ```
You can collapse these into one branch if you want, but you don't need to if you prefer keeping it this way for readability
</blockquote></details>

This is the block that is likely going to grow as we add more conditions, so I'd rather keep all the different rules at the one level of if/else block, so that it grows more naturally
___
*[topside/procedures/tests/test_proclang.py, line 287 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiEZ8o1JpDo1QScX2G:-MHiHM-IEvj1ak4rP8dA:b7qwr4q) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/tests/test_proclang.py#L287)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> compAnd
> ```
Nit: `snake_case`
</blockquote></details>

Fixed in next commit
___
*[topside/procedures/tests/test_proclang.py, line 359 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiEOAD6AYN0Ebw-K-j:-MHiHOVEADL8IQHgbUiZ:bfy85so) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/tests/test_proclang.py#L359)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> set vent_valve to open
> ```
we really want to open the vent valve
</blockquote></details>

Writing new things is hard. I eventually start saying "set s to v" in later tests
___
*[topside/procedures/tests/test_proclang.py, line 363 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiEchd601CuOvz572i:-MHiHWmi3qq969cfboga:b-896fix) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/tests/test_proclang.py#L363)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> compAnd
> ```
Same
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

akmorrison committed:
proc: snake_case in proclang tests

jacobdeery reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-:-MHiIKqNAJCDmZgIK5gO:bnfp4nl)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @akmorrison and @jacobdeery)

___
*[topside/procedures/proclang.py, line 129 at r2](https://reviewable.io/reviews/waterloo-rocketry/topside/66#-MHiE0z9CvRAy68wpsWN:-MHiI1Cv70OOD-8FcGo_:b4xcq8w) ([raw file](https://github.com/waterloo-rocketry/topside/blob/e5cd305a0f54a9d72e6fbce6d2879a1f29f0d48f/topside/procedures/proclang.py#L129)):*
<details><summary><i>Previously, akmorrison (Aaron Morrison) wrote…</i></summary><blockquote>

This is the block that is likely going to grow as we add more conditions, so I'd rather keep all the different rules at the one level of if/else block, so that it grows more naturally
</blockquote></details>

Yep, makes sense


<!-- Sent from Reviewable.io -->
